src/_guides/language/effective-dart/design.md

```diff
@@ -1005,79 +1091,260 @@ var names = people.map((Person person) {
   var value = map[name];
   if (value != null) return value;
   return defaultValue;
 }
 {% endprettify %}
 
 
-### AVOID annotating with `Function`.
+<aside class="alert alert-info" markdown="1">
+
+Before Dart 2, this guideline stated the exact opposite: *don't* annotate with
+`dynamic` when it is implicit. With the new stronger type system and type
+inference, users now expect Dart to behave like an inferred statically-typed
+language. With that mental model, it is an unpleasant surprise to discover that
+a region of code has silently lost all of the safety and performance of static
+types.
+
+</aside>
+
 
-The `Function` type is barely more precise than using no annotation at all. If
-you're bothering to annotate, it's better to use a precise function type that
-describes the signature and return type of the function.
+### AVOID annotating with only `Function`.
 
-If you are annotating a field, this does mean you have to create a typedef, but
-that's usually worth doing.
+The name `Function` by itself without any return type or parameter signature
+refers to the special `Function` type. This type is only marginally more useful
+than using `dynamic`. If you're going to annotate, prefer a full function type
+that describes the signature and return type of the function.
 
 {:.good-style}
 <?code-excerpt "misc/lib/effective_dart/design_good.dart (avoid-Function)"?>
 {% prettify dart %}
-bool isValidString(String value, bool predicate(String string)) => ...
+bool isValid(String value, bool Function(String string) test) => ...
 {% endprettify %}
 
 {:.bad-style}
 <?code-excerpt "misc/lib/effective_dart/design_bad.dart (avoid-Function)"?>
 {% prettify dart %}
-bool isValidString(String value, Function predicate) => ...
+bool isValid(String value, Function test) => ...
+{% endprettify %}
+
+In Dart 1, users sometimes used `Function` to avoid the necessary tedium of
+creating a typedef to use the function type in a field, variable, or generic
+type argument. Dart 2 has a [general function type syntax][fn syntax] that can
+be used anywhere a type annotation can occur, so this shortcut is no longer
+needed.
+
+[fn syntax]: #prefer-inline-function-types-over-typedefs
+
+One exception to this guideline is if you want a type that represents the union
+of multiple different function types. For example, you may accept a function
+that takes one parameter or a function that takes two. Since we don't have union
+types, there's no way to precisely type that and you'd normally have to use
+`dynamic`. `Function` is at least a little more helpful than that:
+
+{:.good-style}
+<?code-excerpt "misc/lib/effective_dart/design_good.dart (function-arity)"?>
+{% prettify dart %}
+void handleError(void Function() operation, Function errorHandler) {
+  try {
+    operation();
+  } catch (err, stack) {
+    if (errorHandler is Function(Object)) {
+      errorHandler(err);
+    } else if (errorHandler is Function(Object, Object)) {
```

Comment:
Cool, I didn't know that we could do that!

Reply:
I think that this example could benefit from some highlighting too: how about using

    <?code-excerpt "..." replace="/Function(\(.*?\))?/[!$&!]/g"?>

Reply:
Done.